docs/network: update network-binding-mechanisms documentation (#17534)
kubevirt/kubevirt: docs/network/network-binding-mechanisms.md

Reviewer: Consider adding a link to https://www.cni.dev/.

Author: I do not know why this is important, but it cannot hurt.
Done.

Reviewer: Consider adjusting this manifest to be more accurate.

Author: Done

> paramount we separate the pod networking configuration and the  VM networking

Reviewer: nit:
```suggestion
paramount we separate the pod networking configuration and the VM networking
```

   includes:
   - IP address
   - Routes (**only** bridge)
   - Gateway (**only** bridge)
   - MAC address (**only** bridge)
   - Link MTU

2. Persist the interface state to the file system:
   - Pod interface cache: records the interface state for virt-handler state
     tracking (applies to all bindings).
   - DHCP interface cache: stores the interface configuration for the in-pod
     DHCP server (applies to bridge binding only).

3. Prepare the pod network interfaces: use the information gathered above to
   perform actions specific to each binding mechanism.
   See each binding mechanism section for more details.

### Unprivileged VMI networking configuration
The virt-launcher is an untrusted component of KubeVirt (since it wraps the
libvirt process that will run third party workloads). As a result, it must be
run with as little privileges as required. The only networking-related
capability required by virt-launcher is `CAP_NET_BIND_SERVICE`, which allows
the in-pod DHCP server to bind to a privileged port.

In this second phase, virt-launcher generates the VM's domain configuration
based on the selected binding mechanism and interface information read from

> netlink. For bridge binding, it additionally starts an in-pod DHCP server

Reviewer: The DHCP server is also started for masquerade (but not using the info from phase 1).

using the configuration written by virt-handler in phase 1.

The domain configuration is specific to each
[binding mechanism](#binding-mechanisms), as each leads to a different
hypervisor configuration.

## Binding Mechanisms
A binding mechanism specifies how to wire the VM network interface to the pod
network interface. It sets up the required networking infrastructure and
generates the hypervisor configuration.

Each interface type has a different binding mechanism, since it leads to a
different hypervisor configuration and may require different networking
infrastructure to be created — e.g. a bridge for the `bridge` or `masquerade`
binding mechanisms.

The existing binding mechanisms are:
- [bridge](#bridge-binding-mechanism)
- [masquerade](#masquerade-binding-mechanism)

### Bridge binding mechanism
Using the bridge binding requires a VMI configuration featuring a
network whose interface type is `bridge` - the yaml file below can be used
as reference, but please refer to the

> [user guide](https://kubevirt.io/user-guide/#/creation/interfaces-and-networks?id=bridge)

Reviewer: This link is broken:
```suggestion
[user guide](https://kubevirt.io/user-guide/network/interfaces_and_networks/#bridge)
```

generates the hypervisor configuration.

Each interface type has a different binding mechanism, since it leads to a
different hypervisor configuration and may require different networking
infrastructure to be created — e.g. a bridge for the `bridge` or `masquerade`
binding mechanisms.

The existing binding mechanisms are:
- [bridge](#bridge-binding-mechanism)
- [masquerade](#masquerade-binding-mechanism)

### Bridge binding mechanism
Using the bridge binding requires a VMI configuration featuring a
network whose interface type is `bridge` - the yaml file below can be used
as reference, but please refer to the
[user guide](https://kubevirt.io/user-guide/network/interfaces_and_networks/#bridge)
for more information.
```yaml
kind: VirtualMachine
spec:
  template:
    spec:
      domain:
        devices:
          interfaces:
            - name: default
              bridge: {}
      networks:
      - name: default
        pod: {}
```

Let's refer to the image below to get a better understanding of how this
binding mechanism works.

Bridge binding mechanism diagram

![Bridge binding mechanism](bridge-binding.svg)

> As can be seen in the diagram above, there are three actors at play: CNI,

Reviewer 2: [nit]: The updated diagram doesnt show the libvirt box anymore, so this line may be a bit confusing?

The virt-launcher is an untrusted component of KubeVirt (since it wraps the
libvirt process that will run third party workloads). As a result, it must be
run with as little privileges as required. As of now, the only capability
required by virt-launcher to configure networking is `CAP_NET_BIND_SERVICE`.

In this second phase, virt-launcher selects the correct
[binding mechanism](#binding-mechanisms) and uses it to retrieve the
configuration data previously gathered in phase 1.

With the interface state loaded, it proceeds to generate the VM's domain
configuration. This generation is specific to each
[binding mechanism](#binding-mechanisms), as each leads to a different
hypervisor configuration.

## Binding Mechanisms
A binding mechanism can be seen as the translation service between KubeVirt's
API and Libvirt's domain xml. It also features methods to set up the required
networking infrastructure.

Each interface type has a different binding mechanism, since it leads to a
different hypervisor configuration and may require different networking
infrastructure to be created — e.g. a bridge for the `bridge` or `masquerade`
binding mechanisms.

> The existing binding mechanisms are:

Author: Makes sense to add SR-IOV and the new passt one, however, I would prefer to do it in follow up in order not to burden this change (it only takes care of existing bindings and removed an old one).

is performed in the *target* virt-launcher's net namespace.

The first action done in the privileged VMI networking configuration is to
identify the correct [binding mechanism](#binding-mechanisms) to use.
Once identified, the following operations are performed, in this order:

1. Discover the pod network interface: gather information about the pod
   interface required by the binding mechanism. The gathered information
   includes:
   - IP address
   - Routes (**only** bridge)

>    - Gateway

Reviewer: Only for bridge binding.

Author: Done

   - MAC address (**only** bridge)
   - Link MTU

2. Prepare the pod network interfaces: use the information gathered above to
   perform actions specific to each binding mechanism.
   See each binding mechanism section for more details.

3. Persist the interface state to the file system, making the configured pod

>    interface available to the virt-launcher pod (phase 2).

Reviewer: Actually, there are two distinct caches written during discovery:
- Pod interface cache: Used by virt-handler for state tracking (applies to all bindings).
- DHCP interface cache: Used by virt-launcher to initialize the DHCP server (applies to bridge binding only).

The current framing suggests that persistence is primarily for making data available to virt-launcher in phase 2, but that logic really only applies to the DHCP cache.

Author: Updated.

### Unprivileged VMI networking configuration
The virt-launcher is an untrusted component of KubeVirt (since it wraps the
libvirt process that will run third party workloads). As a result, it must be
run with as little privileges as required. As of now, the only capability

> required by virt-launcher to configure networking is `CAP_NET_BIND_SERVICE`.

Reviewer: ```suggestion
the only networking-related capability required by virt-launcher is CAP_NET_BIND_SERVICE, which allows the in-pod DHCP server to bind to a privileged port.
```

Author: Done

In this second phase, virt-launcher selects the correct
[binding mechanism](#binding-mechanisms) and uses it to retrieve the

> configuration data previously gathered in phase 1.

Reviewer: virt-launcher uses the files created by virt-handler only to start the DHCP server for bridge binding. The rest of the information is read from netlink.

Author: Done

> With the interface state loaded, it proceeds to generate the VM's domain
> configuration. This generation is specific to each

Reviewer: ```suggestion
The virt-launcher is responsible for generating the VM's domain
configuration. This generation is specific to each
```

Author: Done

[binding mechanism](#binding-mechanisms), as each leads to a different
hypervisor configuration.

## Binding Mechanisms

> A binding mechanism can be seen as the translation service between KubeVirt's

Reviewer: I think we should frame it around bindings closing the gap between the pod and the VM running inside of it.

Author: Rephrased.

configuration step - performed by CNI - is out of scope of this guide. The
focus will be instead on how KubeVirt performs bridge binding.

The bridge binding starts by reading the pod networking interface configured by
CNI. It records the assigned MAC address, MTU, IP address(es), and routes.

In phase 1, KubeVirt creates an in-pod bridge and renames the pod networking
interface (e.g. `eth0` → `eth0-nic`), attaching the renamed interface as a
bridge port. Its MAC address is replaced and its IP addresses are removed.

To preserve the original network identity, a dummy interface is created under
the original pod interface name (e.g. `eth0`). It holds the original MAC
address, IP address(es), MTU, and routes. This serves two purposes:

- The in-pod DHCP server in phase 2 uses it to advertise the correct

>   addressing to the VM, effectively carrying the pod IP into the guest.

Reviewer: AFAIK, the DHCP server only reads the link to get the MTU [1]. The rest is loaded from the file written by virt-handler.

[1] https://github.com/kubevirt/kubevirt/blob/fc289eb919349f6d111c6817d58ad1bd824e250b/pkg/network/dhcp/bridge.go#L66

Author: Right, it was my wishful thinking. This is what I wanted to do as part of the change in the virt-launcher, but it did not realize.
Updated.

configuration step - performed by CNI - is out of scope of this guide. The
focus will be instead on how KubeVirt performs bridge binding.

The bridge binding starts by reading the pod networking interface configured by
CNI. It records the assigned MAC address, MTU, IP address(es), and routes.

In phase 1, KubeVirt creates an in-pod bridge and renames the pod networking
interface (e.g. `eth0` → `eth0-nic`), attaching the renamed interface as a
bridge port. Its MAC address is replaced and its IP addresses are removed.

To preserve the original network identity, a dummy interface is created under
the original pod interface name (e.g. `eth0`). It holds the original MAC

> address, IP address(es), MTU, and routes. This allows the container runtime

Reviewer: The dummy interface doesn't hold routes - it holds the original MAC, IPs, and MTU [1]. Routes are stored separately in the DHCP cache file.

```suggestion
the original pod interface name (e.g. `eth0`). It holds the original MAC
address, IP address(es) and MTU. This allows the container runtime
```

[1] https://github.com/kubevirt/kubevirt/blob/7a4c8b358f3a1e3f764e5b930bc013bf86896703/pkg/network/setup/netpod/netpod.go#L393-L401

to continue answering CNI CHECK commands at runtime: since the IP has been
moved to the guest and removed from the renamed pod interface, the dummy
retains it so CNI can still verify the pod's network configuration is intact.

In phase 2 (executed by virt-launcher, unprivileged) the VM's domain
configuration is generated.

> > **libvirt/QEMU specific:** The following domain XML example applies to the

Reviewer: nit: Consider using the GitHub Note notation:

```suggestion
> [!NOTE]
> **libvirt/QEMU specific:** The following domain XML example applies to the
```

https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#alerts

 </interface>

> ```

Reviewer: Consider dropping this snippet.

Author: Done

</interface>

> ```

Reviewer: This snippet is outdated, the following snippet is more accurate:
```xml
<interface type='ethernet'>
    <target dev='tap0' managed='no'/>
    <model type='virtio'/>
    <mac address='8a:37:e9:71:f2:a4'/>
    <mtu size='1450'/>
    <alias name='ua-default'/>
</interface>
```

Author: Done

Once the VM is booted, libvirt will consume the interface XML definition and

> create a tap device — named after the `target` parameter. That tap device will

Reviewer: Libvirt does not create a tap device. The tap device is created by virt-handler and should already be there when libvirt starts.

Author: Right, I have missed this point.
Done
